lib/semantics/statements.cc

```diff
@@ -0,0 +1,97 @@
+#include "statements.h"
+#include "semantics.h"
+#include "../common/indirection.h"
+#include "../common/Fortran.h"
+#include "../evaluate/expression.h"
+#include "../parser/parse-tree.h"
+#include <optional>
+#include <variant>
+
+namespace Fortran::semantics {
+
+class StatementContext {
+public:
```

Comment:
I don't think you need a context here. (It looks like it merely carries the SemanticsContext, which can be a data member of the checker instance.  See, for example, check-computed-goto.h.)  Aside: there is more state required in the DO CONCURRENT checker since a DO CONCURRENT construct places constraints upon the statements it contains.  That complexity won't happen in a STOP statement, though.  That should let you prune this down, I suspect, to a pair of StopStmtChecker::Leave(...) member functions in your checker.

Reply:
Code rearranged.